flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/util/HBaseSerde.java

```diff
@@ -0,0 +1,371 @@
+	private final FieldDecoder[][] qualifierDecoders;
+
+	public HBaseSerde(HBaseTableSchema hbaseSchema) {
+		this.families = hbaseSchema.getFamilyKeys();
+		this.rowkeyIndex = hbaseSchema.getRowKeyIndex();
+		LogicalType rowkeyType = hbaseSchema.getRowKeyDataType().map(DataType::getLogicalType).orElse(null);
+
+		// field length need take row key into account if it exists.
+		checkArgument(rowkeyIndex != -1 && rowkeyType != null, "row key is not set.");
+		this.fieldLength = families.length + 1;
+
+		// prepare output rows
+		this.reusedRow = new GenericRowData(fieldLength);
+		this.reusedFamilyRows = new GenericRowData[families.length];
+
+		// row key should never be null
+		this.keyEncoder = createFieldEncoder(rowkeyType);
+		this.keyDecoder = createFieldDecoder(rowkeyType);
+
+		this.qualifiers = new byte[families.length][][];
+		this.qualifierEncoders = new FieldEncoder[families.length][];
+		this.qualifierDecoders = new FieldDecoder[families.length][];
+		String[] familyNames = hbaseSchema.getFamilyNames();
+		for (int f = 0; f < families.length; f++) {
+			this.qualifiers[f] = hbaseSchema.getQualifierKeys(familyNames[f]);
+			DataType[] dataTypes = hbaseSchema.getQualifierDataTypes(familyNames[f]);
+			this.qualifierEncoders[f] = Arrays.stream(dataTypes)
+				.map(DataType::getLogicalType)
+				.map(HBaseSerde::createNullableFieldEncoder)
+				.toArray(FieldEncoder[]::new);
+			this.qualifierDecoders[f] = Arrays.stream(dataTypes)
+				.map(DataType::getLogicalType)
+				.map(HBaseSerde::createNullableFieldDecoder)
+				.toArray(FieldDecoder[]::new);
+			this.reusedFamilyRows[f] = new GenericRowData(dataTypes.length);
+		}
+	}
+
+	/**
+	 * Returns an instance of Put that writes record to HBase table.
+	 *
+	 * @return The appropriate instance of Put for this use case.
+	 */
+	public @Nullable Put createPutMutation(RowData row) {
+		byte[] rowkey = keyEncoder.encode(row, rowkeyIndex);
+		if (rowkey.length == 0) {
+			// drop dirty records, rowkey shouldn't be zero length
+			return null;
+		}
+		// upsert
+		Put put = new Put(rowkey);
+		for (int i = 0; i < fieldLength; i++) {
+			if (i != rowkeyIndex) {
+				int f = i > rowkeyIndex ? i - 1 : i;
+				// get family key
+				byte[] familyKey = families[f];
+				RowData familyRow = row.getRow(i, qualifiers[f].length);
+				for (int q = 0; q < this.qualifiers[f].length; q++) {
+					// get quantifier key
+					byte[] qualifier = qualifiers[f][q];
+					// serialize value
+					byte[] value = qualifierEncoders[f][q].encode(familyRow, q);
+					put.addColumn(familyKey, qualifier, value);
+				}
+			}
+		}
+		return put;
+	}
+
+	/**
+	 * Returns an instance of Delete that remove record from HBase table.
+	 *
+	 * @return The appropriate instance of Delete for this use case.
+	 */
+	public @Nullable Delete createDeleteMutation(RowData row) {
+		byte[] rowkey = keyEncoder.encode(row, rowkeyIndex);
+		if (rowkey.length == 0) {
+			// drop dirty records, rowkey shouldn't be zero length
+			return null;
```

Comment:
I don't think we should throw exception here by default. Otherwise, it's meaningless to add this condition. IMO, this is a behavior of the storage (not accept empty key). If there is a need to see the exception, we can add an option in the future.